### PR TITLE
Single call protocol in JS

### PR DESCRIPTION
**Pull request by jacoblee93**
Enables more widespread streaming, batch calls.

### CODE REVIEW

File: langchain/src/schema/runnable.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+      const batchPromises = inputs
+        .slice(i, i + batchSize)
+        .map((input, i) => this.invoke(input, configList[i]));
+      const batchResult = await Promise.all(batchPromises);
+      batchResults.push(batchResult);
+    }
+    return batchResults.flat();
+  }
+
+  async *_createAsyncGenerator(
```
Comment (jacoblee93) [inline langchain/src/schema/runnable.ts:55]: Would love a better name for this if you have suggestions @nfcampos

File: langchain/src/llms/openai-chat.ts [hunk 2]
```diff
@@ -50,11 +50,9 @@ export interface OpenAIChatCallOptions extends OpenAICallOptions {
  * @augments AzureOpenAIChatInput
  */
 export class OpenAIChat
-  extends LLM
+  extends LLM<OpenAIChatCallOptions>
   implements OpenAIChatInput, AzureOpenAIInput
 {
-  declare CallOptions: OpenAIChatCallOptions;
```
Comment (nfcampos) [inline langchain/src/llms/openai-chat.ts:56]: We dont need this anymore?

File: langchain/src/llms/base.ts [hunk 3]
```diff
@@ -63,9 +65,117 @@ export abstract class BaseLLM extends BaseLanguageModel {
...
+      let parsedOptions: CallOptions;
+      if (options?.timeout && !options.signal) {
+        parsedOptions = {
+          ...options,
+          signal: AbortSignal.timeout(options.timeout),
+        };
+      } else {
+        parsedOptions = (options ?? {}) as CallOptions;
+      }
+      delete parsedOptions.tags;
```
Comment (nfcampos) [inline langchain/src/llms/base.ts:126]: we probably want to move this logic to a helper method?

File: langchain/src/llms/openai-chat.ts (hunk 2 above)
Comment (jacoblee93) [reply langchain/src/llms/openai-chat.ts:56]: Ah thanks, good catch

Comment (jacoblee93) [reply langchain/src/llms/openai-chat.ts:56]: Yeah, it's now declared with the type parameter on the superclass here: https://github.com/hwchase17/langchainjs/pull/2106/files/381bc22a2dd83824d3ff53574c78e3711e3d3c54#diff-0498c91d7202afcc70230e285dd4e46dfbd6c2525556c3b1cba7c3ce22e94671R115

Comment (jacoblee93) [reply langchain/src/llms/openai-chat.ts:56]: Screenshot 2023-07-28 at 10 14 00 AM

File: langchain/src/llms/base.ts (hunk 3 above)
Comment (jacoblee93) [reply langchain/src/llms/base.ts:126]: Ah right, will do!

**Pull request merged**